### ISSUE TITLE
Flip the meaning of the `Target::NoBoundsQuery` flag?

### ISSUE THREAD

**Issue by dsharletg** [halide/Halide#3228]
I was just talking with @steven-johnson about the const-correctness issue also raised in #3191. 

I was thinking, a less heavyweight way (than a separate bounds query generated function) is to emit const buffer parameters when there is no bounds query support (setting aside device dirty flags, which I think we are leaning towards moving those to be behind the opaque device pointer). When the bounds query logic is present, we could emit non-const buffer parameters.

This wouldn't be that useful today, but if we flipped the meaning of the flag, so we had a positive `Target::BoundsQuery` flag instead of a negative flag, and the default was no bounds query, then this would be a reasonable behavior.

I think use cases for the bounds query are quite rare (extern stages directly calling other Halide pipelines). Basically, the bounds query code is dead code in the large majority of pipelines (all but one of dozens I'm familiar with).

This would be a bit nice for other reasons too:
- I think I've probably had more issues with confusion due to bounds queries when they aren't necessary than vice-versa (actually needing the bounds queries).
- Reading generated code with bounds queries is slightly annoying, it's a bunch of logic that needs to be figured out and skipped.

**Comment by steven-johnson** [comment]
I like this idea too. (I'd argue that doing this and *also* splitting the bounds query into a separate call might eventually be desirable, but the latter is harder to retrofit and less useful in real-world code that I'm aware of)

**Comment by abadams** [comment]
I'm really not OK with having a const buffer being gifted a device allocation that the user is responsible for cleaning up. C++ rules aside, that's a significant mutation of the buffer that the user needs to be aware of.

**Comment by steven-johnson** [comment]
Disregarding the constness issue, this would still be a win; as Dillon points out, the vast majority of code doesn't actually need the bounds query logic.

**Comment by dsharletg** [comment]
Andrew's point was about the device buffer itself rather than the flags, which is a good one. We could also only mark buffers const if there is no target that is capable of writing device buffers...? That's getting uglier quickly though.

This sort of gets at an unrelated issue that I've always found odd: halide_copy_to_host calls get injected in pipelines that only run CPU code, but only when you have a device target feature flag in your target. It doesn't even have to be the "right" target feature flag either, any will do (as long as some *other* pipeline in your program has the right target feature flag to make the linker happy eventually).

**Comment by steven-johnson** [comment]
> Andrew's point was about the device buffer itself rather than the flags

Right, I get that; I'm asserting that even if we don't attempt to do anything with constness, making the bounds-query code opt-in rather than opt-out seems worthwhile (and AFAICT doesn't have anything to do with device buffers getting created)

**Comment by dsharletg** [comment]
Ah right, I agree with that too.

**Comment by abadams** [comment]
Making it a separate entrypoint seems like a minor improvement to me - then there's no flag at all.

**Comment by zvookin** [comment]
I'll need to think about this, but my initial thought is that while this
may have been a better initial design, I don't think it is worth doing by
itself now. In combination with some other changes, it might be the right
thing. Part of the disconnect here is that I think bounds queries are more
generally useful than some other folks do. E.g. there should likely be a
lot more tests that assert things about bounds queries. Halide has a
general problem in that it is somewhere between very difficult and
impossible to fully specify the bounds inference process. ("Impossible" ->
given a strong enough theorem prover, it likely falls afoul of
incompleteness. In any event, it is hard to give a programmer useful spec
that says what Halide programs do and do not do re: bounds.) Thus writing
tests that assert bounds properties is productive.

I do not like the "host == nullptr" means bounds query thing. It is a very
pragmatic design but seems hacky and does result in surprises as Dillon
points out. We'd talked about adding a flag to request this but I'm not
sure that makes a whole lot of difference. Moving bounds query to a
separate call seems like a better design, but we probably still need a bool
per buffer argument in that case. (Andrew's last email says otherwise, but
doesn't the bounds query algorithm allow partial queries and that is
actually useful? In fact I think it might be necessary in some cases.) The
separate call should allow our end users to not think about this much at
all if linker dead stripping works.

Re: buffers being gifted allocations, we have a higher level problem that
we need to expose the resource usage of a Halide generated pipeline and
give developers control over the whole thing. The main driving use case is
that intermediate buffers in a pipeline are allocated every time the
pipeline is called and this is too expensive. (For cuda malloc at least,
but Amdahl's law probably makes malloc an interesting case too. That is as
we make everything else fast, calls to malloc become more significant.) I
don't have a really fleshed out design, but it seems clear to me we need to
do something that collects allocations/gpu-kernel-compilations/etc. into a
full list of stuff for a pipeline and then allow the user to control it. In
this model, device allocations for outputs could end up attached to this
global state -- e.g. a pool -- and be tracked/freed that way.

-Z-

**Comment by abadams** [comment]
The bounds_query entrypoint would indeed still need a way of knowing which buffers are being queried, so separating the entrypoint is orthogonal to the issue with a bounds query flag vs signalling it using host and device being null

**Comment by abadams** [comment]
... per-buffer bounds query flag, that is - the target bounds query flag being discussed is about enabling or disabling the feature entirely.

**Comment by dsharletg** [comment]
Regarding device allocations, I was thinking further: we could just make it an error if the device allocation doesn't exist when it should be, rather than allocating it automatically. This would mirror the behavior for host allocations.

In practice, whenever I expect a device allocation for an input or an output buffer, I always allocate it manually anyways (because I don't want to rely on automation for this).

**Comment by zvookin** [comment]
For inputs, I'd probably want that situation to turn into allocate/free
inside the pipeline rather than an error.

-Z-

**Comment by jingpu** [comment]
+1 for allocate/free inside the pipeline instead of an error. I think in practice people don't manually allocate device allocation for small buffers (e.g. LUTs) are going to be used in a device launch. This also reminds of an issue we found in a downstream use case that the gifted device buffer of a static buffer is not freed inside the pipeline, so causes one-time leak per process.

**Comment by steven-johnson** [comment]
I'm leaning towards the most extreme variant of @dsharletg comment here: i.e. don't *ever* have the code do device allocations for you under the hood. The auto-allocate-and-copy is convenient but surely real-world pipelines want to know when allocations and copies are being done, and doing them manually shouldn't be an onerous burden (and if they are, surely we can fix that).

**Comment by abadams** [comment]
I still want to cater to lightweight non-Google usage. An option is to move this sort of logic into the JIT layer (query the pipeline to see if it needs a device allocation, and if so make one inside Pipeline.cpp)

**Comment by steven-johnson** [comment]
Catering to lightweight usage is definitely important, but let me push back a little bit and suggest that even for casual usage, you should be aware of any time a buffer is transitioned from host to device or back; requiring explicit code to enable this shouldn't be unreasonable (and arguably makes the cost involved in the code more obvious).

**Comment by zvookin** [comment]
Removing automatic GPU allocation entirely is a nonstarter.

Allocation plans will hopefully be a good step forward, but transitioning buffers to/from the GPU is not particularly problematic at the moment and removing it entirely breaks existing code, makes a lot of things harder, and doesn't make much of anything easier. (And in modern unified memory setups, the transition cost is very low. We should likely add a way to allocate all host memory in the right kind of address space on systems that support such and teach the GPU backends to wrap instead of copy.)

**Comment by dsharletg** [comment]
I don't think we should eliminate automated GPU host <-> device transfers, only automated allocation of output buffers. Input buffers can be automatically allocated, but they should also be freed before returning, so there is no buffer mutability implication there.

It is a messy change because while it's not a breaking change, it would introduce a performance bug in pipelines that currently implicitly repeatedly read a CPU buffer on the GPU. On the other hand, in today's world, the equivalent bug is a memory usage bug, which is pretty bad too.

**Comment by zvookin** [comment]
Input buffers return their allocations as well. This is prevents multiple GPU copies on the same data.

**Comment by abadams** [comment]
I would note that you get the semantics Dillon wants by simply making a local copy of the Buffer object on the stack before entering the Halide pipeline, or equivalently, by passing your input Buffer objects to the immediately containing function by value instead of const-ref.
